### PR TITLE
SPRacing H7 with QSPI

### PR DESCRIPTION
**Pull request by andyp1per**
Seems to all be working

image

I am currently booting via the SSBL - I looked into getting the primordial bootloader to boot our bootloader but that seems like a lot of work, not sure its worth it since if you care you can simply overwrite the primordial bootloader.

### CODE REVIEW

**Comment by davidbitton** [comment]
Are you using the PX4 version of the SPRacing board with the alternative winbond chip?

**Comment by andyp1per** [comment]
> 
> 
> Are you using the PX4 version of the SPRacing board with the alternative winbond chip?

Yes

**Comment by hydra** [comment]
I've subscribed to this PR so I can follow your progress.  If you need any assistance let me know.  Due to other work I don't have a lot of time available, but I'll help where I can.

**Comment by davidbitton** [comment]
@andyp1per i have two PX4 version boards if need to test.

**Comment by mluessi** [comment]
@andyp1per I have started looking into this as well. 

I'm working on a flight controller that is similar to the SPRacing H7, i.e., it uses a STM32 H750 with a NOR flash via QSPI that be used to execute code from. We use a custom bootloader to configure the flash chip, enable the QSPI memory-mapped mode, and then jump to the ardupilot code.

I based my work off a newer version of the master branch (with ChibiOS that supports the H750) with some your changes. The bootloader jumps to the ardupilot code located in QSPI flash and the code is running, but more changes are needed to get it fully working.

What I did so far:
- Added a Python module for the H750 (STM32H750xx.py)
- Made changes similar to yours to chibios_hwdef.py such that the QSPI pins are configured correctly (AF QSPI), but the QSPI configuration itself is not touched. The bootloader already configures it and the code is currently running from QSPI. Code would need to be running from RAM before the QSPI configuration can be changed.
- Commented out `rccResetAHB4(STM32_GPIO_EN_MASK);` in `stm32_gpio_init`, as resetting the QSPI mapped pins causes a fault.

What still needs to be done, as far as I can tell:
- The vector table should be copied to RAM. The H7 programming manual says that using vendor-specific memory areas for it can cause unpredictable behavior.
- High priority code should be copied to ITCM and/or SRAM.
- It would be nice to also use the external flash for parameter storage. To do so, the relevant code would need to be running from RAM. The QSPI memory mapping temporarily disabled and then re-enabled after writing. Initially, the parameters can be stored on the MicroSD card, so this isn't a very high priority.

The copying of the vector table and code could be done in `__early_init` (board.c). But we would need to identify what code to copy and what changes to the linker script are needed. Copying time critical code to the ITCM RAM would also benefit other H7 and F7 based flight controllers.

I'm happy to help with this, but I'm new to the ardupilot code base, so it would be good to figure out how to do it first.

**Comment by andyp1per** [comment]
@mluessi good stuff! Are you able to boot with your changes? I am slightly hamstrung on the spracing because I don't want to overwrite the original bootloader just yet.

**Comment by mluessi** [comment]
@andyp1per yes, it boots, but is currently hanging in `spi_lld_wait_complete` during the sdcard initialization. I assume it's interrupt or DMA related. 

I think it should be possible to make this work with the ardupilot bootloader and other proprietary bootloaders. E.g. the spracing or the BrainFPV one that my target uses. As long as the bootloader configures the flash chip and enables the QSPI memory mapped mode, it should work.

Are you able to jump to the firmware and debug with the SPRacing bootloader?

**Comment by andyp1per** [comment]
I've not got as far as something that boots and not yet tried wiring up the debugger. I was waiting for the ChibiOS 20.3 merge and then got waylaid with other projects.

**Comment by mluessi** [comment]
I see. The ChibiOS changes were the reason I started a new branch instead of using this branch directly. 

Any ideas on how to proceed with this?

**Comment by andyp1per** [comment]
I've re-based this branch on latest 20.3. Do you have a link to your PR?

**Comment by mluessi** [comment]
@andyp1per I haven't made a PR yet. I will re-create my changes based on your updated branch and then make a PR

**Comment by andyp1per** [comment]
Or do a PR against this PR and I can merge - I don't think we'll have anything mergeable to master any time soon. I could also make this branch writable by you so that we can collaborate.

**Comment by andyp1per** [comment]
Rebased

**Comment by mluessi** [comment]
@andyp1per please excuse the long silence. I just opened a PR against this: https://github.com/andyp1per/ardupilot/pull/26 It adds a module for the H750 and initial support for the RADIX 2 HD, which is similar to the SPRacing H7, i.e., it uses a H750 with code running from QUADSPI.

I think the main work for this to work will be to identity what code needs to be copied to RAM and how to best do so. I assume the copying can be done in `__early_init`. Copying critical code to TCM RAN will also benefit H7 flight controllers that run code from flash.

File: libraries/AP_HAL_ChibiOS/hwdef/SPRacingH7/hwdef.dat [hunk 1]
```diff
@@ -0,0 +1,169 @@
+# hw definition file for processing by chibios_pins.py
+# for SP/Racing Extreme H7 hardware.
+# thanks to betaflight for pin information
+
+# MCU class and specific type
+MCU STM32H7xx STM32H743xx
```
Comment (davidbitton) [inline libraries/AP_HAL_ChibiOS/hwdef/SPRacingH7/hwdef.dat:6]: Why is the MCU spec'ed as H743 and not H750

Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/SPRacingH7/hwdef.dat:6]: @mluessi only just provided a definition script, this was a shortcut by me before that was available. I'll fix when I get back to looking at this.

**Comment by andyp1per** [comment]
Ok taking a different tack here. I have a NucleoH743 and a QSPI daughterboard that I am going to connect. I have modified the config so that it works ok on the Nucleo, will be easier to debug the QSPI issues that way

**Comment by peterbarker** [comment]
@andyp1per does this need a rebase?

**Comment by andyp1per** [comment]
So thanks to @bugobliterator 's hard work this is mostly working. The big remaining issue is parameter storage which in theory should fall back to the SD card but the SD card does not appear to be working.

**Comment by andyp1per** [comment]
@hydra any top tips for running the debugger with the spracing? I can't seem to set breakpoints because of the external flash:

```
(gdb) cont
Continuing.
Warning:
Cannot insert breakpoint 4.
Cannot access memory at address 0x901b1bb8
Cannot insert breakpoint 2.
Cannot access memory at address 0x901b4ca8
Cannot insert breakpoint 1.
Cannot access memory at address 0x901b4d4c
```
is there a special openocd config file I need?

**Comment by hydra** [comment]
> @hydra any top tips for running the debugger with the spracing? I can't seem to set breakpoints because of the external flash:

Obviously, you need to make sure you've loaded the symbol files in GDB and made sure you've compiled with debugging information on, I've had better results with gdawrf5 symbols. (IIRC).

I've had mixed results when working with projects that use multiple symbol files.  Some debuggers seem to work better than others, I usually use Black Magic Proble and J-Link, I've had best results with J-Link.  OpenOCD has always been a bit sub-par for me.  It does work though.  Sometimes I find that setting a breakpoint after you have paused the target works better than before it's paused.  Try setting the debugger to connect and reset before it uploads the list of breakpoints to the target.  I've also found that in eclipse disabling all breakpoints (via the toolbar button), then connecting, then enabling them all worked.

Suggest getting it working using the command-line GDB interface first as you can determine the order of things easier that way, then work around any ide quirks.

Also, it's possible to use Segger STLink Reflash to convert an stlink debugger into a J-Link (and you can reverse the process too).

**Comment by hydra** [comment]
> @hydra any top tips for running the debugger with the spracing? I can't seem to set breakpoints because of the external flash:

> Cannot insert breakpoint 4.
> Cannot access memory at address 0x901b1bb8

also... this looks like your debugger is attempting to access the quad/octo spi region of ram, which is doesn't actually need to do to _set_ a _hardware_ breakpoint.  is your debugger trying to insert 'bpkt' instructions?  if so that will obviously never work because a) the memory region isn't enabled at reset, and b) it's read-only anyway.  Ensure it's setting hardware breakpoints, no software breakpoints.

**Comment by andyp1per** [comment]
@hydra - thanks, I made a few changes to my gdb and openocf configs and now things seem to be working - thanks!

**Comment by andyp1per** [comment]
Giving up on this for now, the flash on this board is just too slow to be practical for our architecture

**Comment by andyp1per** [comment]
I've not quite given up. Basically anything that is performance critical has to go in RAM. Tracking those things down is a painstaking task of putting timing measurements in code and incrementally changing the linker script. Even then to get to something safe to fly would require building this into a copter and doing the same again. A long, painful road which I am rapidly getting disillusioned with.

**Comment by andyp1per** [comment]
With the current changes I end up with about 10% idle cycles. The ArduCopter thread is 43% (should be 20%). Might be enough to fly, but who knows.

**Comment by hydra** [comment]
It's well worth continuing with the process.  You will undoubtedly find a host of areas in the AP codebase that can be optimised and that will benefit all the AP users.  Its likely worth improving the methods used to profile the code so that future optimization is also easier to achieve.  Additionally its worth also checking the compiler and linker code generation and optimization settings and making aure you're not missing out on potential optimizations.  You have made excellent progress so far!

**Comment by andyp1per** [comment]
Thanks @hydra - I'm on to my next project now. Will pick this up again once the days get shorter and flying is a luxury.

I have already tried a whole bunch of different optimizations and tricks and while these make some difference it's only about 20% max. The problem is that running direct from external flash is 200% - 800% slower than on a regular H743, so 20% is chump change in that context. The only place where performance is reasonable is in RAM and so this is now a question of which critical bits do I need to put in RAM and can I get enough in RAM without running out 😄

**Comment by hydra** [comment]
> 200% - 800% slower

200-800% slower isn't what ST say in their docs.

https://www.st.com/resource/en/application_note/an5188-external-memory-code-execution-on-stm32f7x0-value-line-stm32h750-value-line-stm32h7b0-value-line-and-stm32h730-value-line-mcus-stmicroelectronics.pdf

https://www.st.com/en/embedded-software/x-cube-perf-h7.html

File: Tools/CPUInfo/CPUInfo.cpp [hunk 2]
```diff
@@ -28,8 +31,13 @@ static uint32_t sysclk = 0;
 static EKF_Maths ekf;
 
 HAL_Semaphore sem;
+AP_ESC_Telem telem;
 
 void setup() {
+#if defined(STM32H7)
```
Comment (tridge) [inline Tools/CPUInfo/CPUInfo.cpp:37]: we'd need to make this optional with a #define, and not change cache settings by default

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 3]
```diff
@@ -162,6 +162,10 @@ void AP_FlashIface_JEDEC::reset_device()
 {
     // Get chip out of XIP mode
     AP_HAL::QSPIDevice::CommandHeader cmd;
+#ifndef HAL_BOOTLOADER_BUILD
+    _dev->get_semaphore()->take_blocking();
```
Comment (tridge) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:166]: need a comment here. why never released? why not in bootloader?

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp [hunk 4]
```diff
@@ -463,6 +464,7 @@ void Scheduler::_monitor_thread(void *arg)
 void Scheduler::_rcin_thread(void *arg)
 {
     Scheduler *sched = (Scheduler *)arg;
+    ExpandingString str;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Scheduler.cpp:467]: what is this for?

File: libraries/AP_HAL_ChibiOS/Storage.cpp [hunk 5]
```diff
@@ -442,6 +442,12 @@ bool Storage::_flash_erase_ok(void)
  */
 bool Storage::healthy(void)
 {
+#ifdef USE_POSIX
+    // SD card storage is really slow
+    if (_initialisedType == StorageBackend::SDCard) {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Storage.cpp:447]: likely could make this better by increasing CH_STORAGE_LINE_SHIFT, try a value of 9 for 512 byte sectors

File: libraries/AP_HAL/AP_HAL_Boards.h [hunk 6]
```diff
@@ -270,7 +270,7 @@
 #endif
 
 #ifndef HAL_WITH_MCU_MONITORING
-#define HAL_WITH_MCU_MONITORING defined(STM32H7)
+#define HAL_WITH_MCU_MONITORING defined(STM32H7) && !defined(STM32H750xx)
```
Comment (tridge) [inline libraries/AP_HAL/AP_HAL_Boards.h:273]: use STM32_HAS_ADC3

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf.ld [hunk 7]
```diff
@@ -72,18 +87,107 @@ SECTIONS
...
+    .fastramfunc : ALIGN(4) SUBALIGN(4)
     {
         . = ALIGN(4);
-        __instram_init_text__ = LOADADDR(.ramfunc);
+        __instram_init_text__ = LOADADDR(.fastramfunc);
         __instram_init__ = .;
         EXCLUDE_FILE (*vectors.o *crt0_v7m.o *crt1.o)
-        *libch.a:*(.text .text.* .rodata .rodata.* .glue_7t .glue_7 .gcc*)
-        *(.ramfunc)
+        *libch.a:ch*.*(.text* .rodata*)
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/common_extf.ld:97]: comments would be good

**Review by tridge — CHANGES_REQUESTED** [review]
try -Os as default, same as we do for 1M boards, to make code more compact

**Comment by tridge** [comment]
use Tools/scripts/configure_all.py to see if any hwdef.h output changes

File: libraries/AP_HAL_ChibiOS/hwdef/SPRacingH7/hwdef.dat (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/SPRacingH7/hwdef.dat]: Fixed

File: Tools/CPUInfo/CPUInfo.cpp (hunk 2 above)
Comment (andyp1per) [reply Tools/CPUInfo/CPUInfo.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp (hunk 4 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/Scheduler.cpp]: Removed

File: libraries/AP_HAL_ChibiOS/Storage.cpp (hunk 5 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/Storage.cpp:447]: Changed

File: libraries/AP_HAL/AP_HAL_Boards.h (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_HAL/AP_HAL_Boards.h]: Fixed

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf.ld (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf.ld]: Comments added

**Comment by andyp1per** [comment]
-Os loses about 2% performance overall

**Comment by andyp1per** [comment]
-O3 is no better than -O2

**Comment by andyp1per** [comment]
dirdiff checked

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:166]: Needed for the the test, comment added.

File: libraries/AP_HAL_ChibiOS/Storage.h [hunk 8]
```diff
@@ -29,7 +29,13 @@
 
 // when using flash storage we use a small line size to make storage
 // compact and minimise the number of erase cycles needed
+#ifdef STORAGE_FLASH_PAGE
+#define CH_STORAGE_LINE_SHIFT 3
+#elif defined(USE_POSIX)
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Storage.h:34]: needs to be defined(USE_POSIX) && !defined(HAL_WITH_RAMTRON)

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py [hunk 9]
```diff
@@ -905,21 +909,29 @@ def write_mcu_config(f):
...
     env_vars['BOARD_FLASH_SIZE'] = flash_size
-    f.write('#define CRT1_AREAS_NUMBER 1\n')
 
     flash_reserve_start = get_config(
         'FLASH_RESERVE_START_KB', default=16, type=int)
     f.write('\n// location of loaded firmware\n')
     f.write('#define FLASH_LOAD_ADDRESS 0x%08x\n' % (0x08000000 + flash_reserve_start*1024))
-    f.write('#define EXTERNAL_PROG_FLASH_MB %u\n' % get_config('EXTERNAL_PROG_FLASH_MB', default=0, type=int))
+    # can be no persistent parameters if no space allocated for them
+    if not args.bootloader and flash_reserve_start == 0:
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:918]: why disable this? should work fine

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/Storage.h (hunk 8 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/Storage.h]: Fixed

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py (hunk 9 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:918]: This is only disabling it if there is no space reserved

**Review by tridge — APPROVED** [review]

**Pull request merged**